src/OrchardCore/OrchardCore/Modules/Extensions/ServiceCollectionExtensions.cs

```diff
@@ -157,20 +156,23 @@ private static void AddAntiForgery(OrchardCoreBuilder builder)
             builder.ConfigureServices((services, serviceProvider) =>
             {
                 var settings = serviceProvider.GetRequiredService<ShellSettings>();
-                var httpContextAccessor = serviceProvider.GetRequiredService<IHttpContextAccessor>();
 
                 var tenantName = settings.Name;
-                var tenantPrefix = "/" + settings.RequestUrlPrefix;
 
-                // 'PathBase' doesn't include any tenant prefix when called from 'ConfigureServices()'.
-                var pathBase = httpContextAccessor.HttpContext?.Request.PathBase ?? new PathString(String.Empty);
+                // Re-register the antiforgery  services to be tenant-aware.
+                var collection = new ServiceCollection()
+                    .AddAntiforgery(options =>
```

Comment:
Looks like we had to adopt this pattern in a few places already (data protection, authentication and now antiforgery). Would it be worth reconsidering how tenant services are registered? Would it be sensible to initially inject an empty `IServiceCollection` in the module startups and then, only add the host services to the tenant services collection if no service type of the same type was registered by a module?

Not saying it's a better approach, just something that could be explored/debated.

Reply:
Good idea, i will check if there is no side effects, e.g if there is some places where we really want to add a service only if it is not already registered, even it has been registered at the host level.

Reply:
A much more radical approach would be to consider that tenants have their own services that are completely separated from the host services (i.e tenant services wouldn't inherit from the services defined by the host). That's pretty much what EF Core - that maintains an internal `IServiceProvider` for each `DbContext` - does.

Reply:
So that a tenant becomes a fully isolated "application". Hmm, i'm just thinking about all the services registered e.g by the `WebHostBuilder`.

Reply:
Yeah, if we went that far, we'd likely have to "copy" some common services, like `IHostingEnvironment`, `IConfiguration`, `ILoggerFactory`, `IApplicationLifetime`, `DiagnosticSource`/`DiagnosticListener`. We could also have a `OrchardCoreBuilder.PreserveHostService<TService>()` helper allowing to "transfer" host services.

Reply:
Actually, this reminds me of a recent discussion where I was wondering if it wasn't better if OrchardCore was no longer integrated into ASP.NET Core's Startup class. Instead, it could its own "host", that you would configure using the generic host builder (which will replace `WebHost` in 3.0):

```csharp
public static void Main(string[] args)
{
    CreateHostBuilder(args).Build().Run();
}

public static IHostBuilder CreateHostBuilder(string[] args) =>
    Host.CreateDefaultBuilder(args)
        // Register an OrchardCore ASP.NET Core pipeline:
        .ConfigureOrchardCoreWebHost(builder =>
        {
            // ...
        })

        // You can also configure a regular ASP.NET Core 3.0 app if you want to.
        .ConfigureWebHostDefaults(builder =>
        {
            builder.UseStartup<Startup>();
        });
```

At first, it might seem harsh to hide all the host `IApplicationBuilder` machinery, but in the context of OC, it would probably be a good thing (at least, IMHO).